apiserver/facades/agent/uniter/storage.go

```diff
@@ -182,17 +184,17 @@ func (s *StorageAPI) getOneStateStorageAttachment(canAccess common.AuthFunc, id
 }
 
 func (s *StorageAPI) fromStateStorageAttachment(ctx context.Context, stateStorageAttachment state.StorageAttachment) (params.StorageAttachment, error) {
+	// hostTag is the tag either of the unit (this is the case in K8s), or the
+	// machine if the unit is assigned.
 	var hostTag names.Tag
 	hostTag = stateStorageAttachment.Unit()
-	u, err := s.backend.Unit(hostTag.Id())
-	if err != nil {
+	// If the unit is not assigned, we keep the hostTag as initially defined
+	// (i.e. the unit tag).
+	machineName, err := s.applicationService.GetUnitMachineName(ctx, unit.Name(hostTag.Id()))
+	if err != nil && !errors.Is(err, applicationerrors.UnitMachineNotAssigned) {
 		return params.StorageAttachment{}, err
-	}
-	if u.ShouldBeAssigned() {
-		hostTag, err = unitAssignedMachine(s.backend, stateStorageAttachment.Unit())
-		if err != nil {
-			return params.StorageAttachment{}, err
-		}
+	} else if err == nil {
+		hostTag = names.NewMachineTag(machineName.String())
 	}
```

Comment:
This is not nice, but I assume it's necessary given `StorageAttachmentInfo` hasn't been re-written for DQLite yet?

When storage is done, we should loop back around and improve this